lib/ttlcache/eviction.go

```diff
@@ -0,0 +1,195 @@
+// it is nil, or because it was removed.
+const NotIndexed = -1
+
+// Index returns the index of this entry within the heap.
+func (e *Entry) Index() int {
+	if e == nil {
+		return NotIndexed
+	}
+	return e.heapIndex
+}
+
+// Key returns the key for the entry in the heap.
+func (e *Entry) Key() string {
+	return e.key
+}
+
+// ExpiryHeap is a heap that is ordered by the expiry time of entries. It may
+// be used by a cache or storage to expiry items after a TTL.
+//
+// ExpiryHeap expects the caller to synchronize calls to most of its methods. This
+// is necessary because the cache needs to ensure that updates to both its
+// storage and the ExpiryHeap are synchronized.
+type ExpiryHeap struct {
+	entries []*Entry
+
+	// NotifyCh is sent a value whenever the 0 index value of the heap
+	// changes. This can be used to detect when the earliest value
+	// changes.
+	NotifyCh chan struct{}
+}
+
+// NewExpiryHeap creates and returns a new ExpiryHeap.
+func NewExpiryHeap() *ExpiryHeap {
+	h := &ExpiryHeap{NotifyCh: make(chan struct{}, 1)}
+	heap.Init((*entryHeap)(h))
+	return h
+}
+
+// Add an entry to the heap.
+//
+// Must be synchronized by the caller.
+func (h *ExpiryHeap) Add(key string, expiry time.Duration) *Entry {
+	entry := &Entry{
+		key:    key,
+		expiry: time.Now().Add(expiry),
+		// Set the initial heap index to the last index. If the entry is swapped it
+		// will have the correct index set, and if it remains at the end the last
+		// index will be correct.
+		heapIndex: len(h.entries),
+	}
+	heap.Push((*entryHeap)(h), entry)
+	if entry.heapIndex == 0 {
+		h.notify()
+	}
+	return entry
+}
+
+// Update the entry that is currently at idx with the new expiry time. The heap
+// will be rebalanced after the entry is updated.
+//
+// Must be synchronized by the caller.
+func (h *ExpiryHeap) Update(idx int, expiry time.Duration) {
+	if idx == NotIndexed {
+		// the previous entry did not have a valid index, its not in the heap
+		return
+	}
+	entry := h.entries[idx]
+	entry.expiry = time.Now().Add(expiry)
+	heap.Fix((*entryHeap)(h), idx)
+
+	// If the previous index and current index are both zero then Fix did not
+	// swap the entry, and notify must be called here.
+	if idx == 0 || entry.heapIndex == 0 {
+		h.notify()
+	}
+}
+
+// Remove the entry at idx from the heap.
+//
+// Must be synchronized by the caller.
+func (h *ExpiryHeap) Remove(idx int) {
+	entry := h.entries[idx]
+	heap.Remove((*entryHeap)(h), idx)
+
+	// A goroutine which is fetching a new value will have a reference to this
+	// entry. When it re-acquires the lock it needs to be informed that
+	// the entry was expired while it was fetching. Setting heapIndex to -1
+	// indicates that the entry is no longer in the heap, and must be re-added.
+	entry.heapIndex = NotIndexed
+
+	if idx == 0 {
+		h.notify()
+	}
+}
+
+type entryHeap ExpiryHeap
+
+func (h *entryHeap) Len() int { return len(h.entries) }
```

Comment:
I moved the `heap.Interface` methods onto a new unexported type so that they wouldn't distract from the methods that the caller should be using.